File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    private static final int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    public int numIslands(char[][] grid) {
        if (grid.length == 0){
            return 0;
        }
        int rows = grid.length;
        int cols = grid[0].length;
        int islands = 0;
        for (int i = 0; i < rows; i ++){
            for (int j = 0; j < cols; j ++){
                if (grid[i][j] == '1'){
                    backTrack(grid, i, j);
                    islands ++;
                }
            }
        }
        return islands;
    }

    private void backTrack(char[][] grid, int x, int y){
        Queue<int[]> visit = new LinkedList<>();
        grid[x][y] = '0';
        visit.add(new int[]{x, y});
        while (!visit.isEmpty()){
            int[] current = visit.poll();
            int row = current[0], col = current[1];
            for (int i = 0; i < directions.length; i ++){
                int nr = row + directions[i][0], nc = col + directions[i][1];
                if (nr >= 0 && nc >= 0 && nr < grid.length && nc < grid[0].length && grid[nr][nc] == '1'){
                    visit.add(new int[]{nr, nc});
                    grid[nr][nc] = '0';
                }
            }
        }
    }
}

File: Data Structures & Algorithms/number-of-distinct-islands/submission-0.py
class Solution:
    def numDistinctIslands(self, grid: List[List[int]]) -> int:
        if not grid:
            return 0

        seen_shapes = set()
        directions = [(-1, 0), (1, 0), (0, 1), (0, -1)]
        ROWS, COLS = len(grid), len(grid[0])

        #DFS algo
        curr_hash = []
        def dfs(s_row, s_col, row, col):
            #get hash number relative to s_row and s_col
            diff_r = s_row - row
            diff_c = s_col - col
            hash_val = diff_r * COLS + diff_c
            curr_hash.append(hash_val)
            grid[row][col] = 0
            for dr, dc in directions:
                new_r, new_c = row + dr, col + dc
                if (0 <= new_r < ROWS) and (0 <= new_c < COLS) and grid[new_r][new_c] == 1:
                    dfs(s_row, s_col, new_r, new_c)
        
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1:
                    curr_hash = []
                    dfs(r, c, r, c)
                    seen_shapes.add(tuple(curr_hash))
        
        return len(seen_shapes)